### ISSUE TITLE
Error when deleting an article that has one or more comments

### ISSUE THREAD

**Issue by yourguessismine** [HDInnovations/UNIT3D#439]
### Issue Type
- [x] Bug Report
- [ ] Feature Idea
- [ ] Enhancement
- [ ] Technical Discussion
- [ ] Question

When deleting an article that contains one or more comments a 500 error is returned. Deleting an article that has no comments deletes just fine.

### End User Info
-- UNIT3D Version: 1.8.6

-- Laravel Version: 5.6.39

-- PHP Version: 7.2

-- Database Driver & Version: MySQL 5.7.23

-- Web Server Driver & Version: NGINX 1.15.0

-- OS Driver and Version: Ubuntu 18.04.1

### Expected Behavior:

Article deletes along with any comments.

### Actual Behavior:

500 Error:

`SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`HD4Free`.`comments`, CONSTRAINT `fk_comments_articles_1` FOREIGN KEY (`article_id`) REFERENCES `articles` (`id`) ON UPDATE CASCADE) (SQL: delete from `articles` where `id` = 15) {"userId":3,"email":"cdouble@hd4.xyz","exception":"[object] (Illuminate\\Database\\QueryException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`HD4Free`.`comments`, CONSTRAINT `fk_comments_articles_1` FOREIGN KEY (`article_id`) REFERENCES `articles` (`id`) ON UPDATE CASCADE) (SQL: delete from `articles` where `id` = 15) at /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connection.php:664, Doctrine\\DBAL\\Driver\\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`HD4Free`.`comments`, CONSTRAINT `fk_comments_articles_1` FOREIGN KEY (`article_id`) REFERENCES `articles` (`id`) ON UPDATE CASCADE) at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:144, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`HD4Free`.`comments`, CONSTRAINT `fk_comments_articles_1` FOREIGN KEY (`article_id`) REFERENCES `articles` (`id`) ON UPDATE CASCADE) at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:142)`

### Steps To Reproduce The Problem:
1. Create article
2. Comment on article
3. Delete article

### Possible Fix:

?

**Comment by artyuum** [comment]
**Possible Fix:**
Automatically deleting article comments using [Eloquent "deleting" event](https://laravel.com/docs/5.7/eloquent#events)

```php
// App/Article.php

/**
* Automatically deletes article comments.
*/
public static function boot() {
    parent::boot();

    static::deleting(function($article) {
        $article->comments()->delete();
    });
}
```
Tested in my local env and it worked.

@HDVinnie If you agree with this fix I'll make a commit.

**Comment by artyuum** [comment]
@HDVinnie I believe that this kind of task should remain in the Model as this is related to the database. You can see this task like `ON CASCADE DELETE`. Keeping this in the Model will also allow us to lighten the controllers.

**Comment by HDVinnie** [comment]
@artyuum Im fine with that. However in this case I think we should make a new branch and convert any controllers that have a `delete` function to make use of Eloquent model events like the snippet you posted above. Then make a PR from the branch for the team to review? This way where knocking it all out at once? Agree?

**Comment by HDVinnie** [comment]
After thinking on this more. It might be more "correct" to use Observers.

https://laravel.com/docs/5.7/eloquent#observers

```
    /**
     * Handle the User "deleted" event.
     *
     * @param  \App\Article  $article
     * @return void
     */
    public function deleted(Article $article)
    {
        // Place all relations to delete here
        // $article->comments()->delete();
    }
```

**Comment by JackPriceBurns** [comment]
I know I am a little late to this, but surely this should be a migration?

You can setup the database to delete all the comments on delete like @artyuum mentioned
This is then handled by the database and not some observer, I don't know if any code has been merged in for this yet.

`$table->foreign('article_id')->references('id')->on('articles')->onDelete('cascade');`

**Comment by HDVinnie** [comment]
Correct. The Migrations need refactoring with proper indexing and foreign keys.
